This code demonstrates how to normalize the rows of a 2D NumPy array so that each row has a unit (length 1) Euclidean norm. 

First, a 2x3 array `x` is created. The code then computes the Euclidean norm (also known as the L2 norm) of each row using `np.linalg.norm(x, axis=1, keepdims=True)`. The `axis=1` argument tells NumPy to compute the norm across columns for each row, and `keepdims=True` ensures the result is a column vector, making it compatible for broadcasting during division.

Next, each row of `x` is divided by its corresponding norm, resulting in `x_normalized`. This operation scales each row so that its length (norm) becomes 1, effectively projecting each row onto the unit sphere in 3D space.

The code then prints the original norms and the normalized array. Finally, it verifies the normalization by recalculating the norm of each row in `x_normalized` and printing the result. These should all be 1 (or very close, due to floating-point precision), confirming that each row has been properly normalized. 

A common 'gotcha' is forgetting to use `keepdims=True`, which can lead to shape mismatches during division.

In [ ]:
import numpy as np

x = np.array([[0, 3, 4],
              [2, 6, 4]])

norms = np.linalg.norm(x, axis=1, keepdims=True)
x_normalized = x / norms

print("x: \n", x)
print("norms:\n", norms)
print("x_normalized:\n", x_normalized)

# Calculate and print the norm of each normalized row
row_norms = np.linalg.norm(x_normalized, axis=1)
print("Norm of each row in x_normalized:\n", row_norms)


x: 
 [[0 3 4]
 [2 6 4]]
norms:
 [[5.        ]
 [7.48331477]]
x_normalized:
 [[0.         0.6        0.8       ]
 [0.26726124 0.80178373 0.53452248]]
Norm of each row in x_normalized:
 [1. 1.]


linalg stands for "linear algebra." In NumPy, the linalg module provides functions for performing linear algebra operations, such as computing norms, matrix multiplication, eigenvalues, and solving linear systems. For example, np.linalg.norm calculates vector or matrix norms using this module.


---

## **Row Vector Normalization with NumPy**

This code normalizes each row of a 2D NumPy array so that each row becomes a unit vector (i.e., its Euclidean norm is 1).

### **Given matrix**

Let

$$
\mathbf{X} = \begin{bmatrix}
0 & 3 & 4 \\
2 & 6 & 4
\end{bmatrix}
$$

### **Step 1: Compute the norm of each row**

The norm (length) of row $i$ is calculated as:

$$
\|\mathbf{x}_i\| = \sqrt{\sum_{j} x_{ij}^2}
$$

where $x_{ij}$ is the element in row $i$, column $j$.

For example,

* Row 1: $\sqrt{0^2 + 3^2 + 4^2} = \sqrt{0 + 9 + 16} = \sqrt{25} = 5$
* Row 2: $\sqrt{2^2 + 6^2 + 4^2} = \sqrt{4 + 36 + 16} = \sqrt{56} \approx 7.483$

### **Step 2: Normalize each row**

Each element in row $i$ is divided by the row's norm:

$$
\mathbf{\hat{x}}_i = \frac{\mathbf{x}_i}{\|\mathbf{x}_i\|}
$$

So the normalized matrix is:

$$
\mathbf{X}_{\text{normalized}} = 
\begin{bmatrix}
\frac{0}{5} & \frac{3}{5} & \frac{4}{5} \\
\frac{2}{\sqrt{56}} & \frac{6}{\sqrt{56}} & \frac{4}{\sqrt{56}}
\end{bmatrix}
$$

### **Step 3: Check norms of normalized rows**

The norm of each row in the normalized matrix should be 1:

$$
\left\|\mathbf{\hat{x}}_i\right\| = 1
$$

---

**In summary:**

* Compute row-wise norms.
* Normalize each row by dividing by its norm.
* Verify that each normalized row has unit length.

---

Let me know if you want an alternate version or further explanation!


Absolutely! Here’s a breakdown of the equation:

---

### **Equation**

$$
\|\mathbf{x}_i\| = \sqrt{\sum_{j} x_{ij}^2}
$$

---

### **Explanation**

This equation calculates the **Euclidean norm** (also called the **L2 norm** or **length**) of the $i$th row vector, $\mathbf{x}_i$, in a matrix.

Let’s break down each part:

---

#### **1. $\mathbf{x}_i$**

* Represents the $i$th row vector of a matrix $\mathbf{X}$.
* For example, if
  $\mathbf{X} = \begin{bmatrix} 0 & 3 & 4 \\ 2 & 6 & 4 \end{bmatrix}$,
  then $\mathbf{x}_1 = [0, 3, 4]$, $\mathbf{x}_2 = [2, 6, 4]$.

---

#### **2. $x_{ij}$**

* Represents the element in row $i$, column $j$ of the matrix.
* In $\mathbf{x}_1 = [0, 3, 4]$,
  $x_{11} = 0$, $x_{12} = 3$, $x_{13} = 4$.

---

#### **3. $\sum_j x_{ij}^2$**

* This means:
  **Sum the squares of every element $j$ in row $i$**
* For $\mathbf{x}_1 = [0, 3, 4]$:
  $(0)^2 + (3)^2 + (4)^2 = 0 + 9 + 16 = 25$

---

#### **4. $\sqrt{\ \cdot\ }$**

* Take the square root of the sum above.
* For our example:
  $\sqrt{25} = 5$

---

### **Putting it together**

For any row vector $\mathbf{x}_i$:

* **Square** each element.
* **Sum** the squared values.
* **Take the square root** of that sum.

This gives you the **Euclidean norm** (length) of that row.

---

**In plain English:**

> The norm of a vector is how long it is, found by squaring its entries, adding them up, and taking the square root.

---

If you want a concrete numerical example or a geometric explanation, just ask!
